packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -84,9 +84,18 @@ class ObjectKey extends Key {
 
 typedef void GlobalKeyRemoveListener(GlobalKey key);
 
-/// A GlobalKey is one that must be unique across the entire application. It is
-/// used by widgets that need to communicate with other widgets across the
-/// application's element tree.
+/// A GlobalKey is a [Key] that must be unique across the entire application.
+///
+/// GlobalKeys are useful because widgets that have global keys reparent their
+/// subtrees when they are moved from one location in the tree to another
+/// location in the tree. In order to reparent its subtree, a widget must arrive
+/// at its new location in the tree in the same animation frame in which it was
+/// removed from its old location in the tree.
+///
+/// Because GlobalKeys identify unique subtrees, the GlobalKey object provides
+/// access to other objects that are associated with that subtree, such as the
+/// [BuildContext] for that subtree and, for [StatefulWidget]s, the [State] for
+/// that subtree.
```

Comment:
I would add something about global keys being expensive and suggesting that unless either of the features described here are necessary, to instead consider Key, ValueKeys, or UniqueKeys.